### PR TITLE
Subset before averaging

### PR DESCRIPTION
**Pull request by huard**
This PR tries to reduce memory usage (and network transfers) when performing an `average_shape` operation. 

The idea is to first slice the dataset, then running the spatial averaging code. The underlying hypothesis is that even though the weights are sparse, the calls to get the data may not be smart enough to request only the needed grid points, and we end up having to load the entire grid in memory, even though a region may only cover a few grid cells. To be honest, I didn't know exactly how to measure the size of the data requests, so I'm not 100% sure about this. 

### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes issue #188
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)
- [x] HISTORY.rst has been updated (with summary of main changes)
- [x] I have added my relevant user information to `AUTHORS.md`

* **What kind of change does this PR introduce?:** 
Optimization

* **Does this PR introduce a breaking change?:** 
No

* **Other information:** 
I'm looking for ideas on how to benchmark this to confirm that it does reduce the memory usage and network requests (when used on an OPENDAP data source). 
@aulemahal

### CODE REVIEW

**Comment by aulemahal** [comment]
I can confirm that xESMF's internal will load the whole grid. It uses `xr.apply_ufunc` with the spatial dimensions as "core dimensions" and thus, the regridding function receives the whole plane as numpy array and _then_ multiplies by the (sparse) weights. Also the use of `xr.apply_ufunc` forbids dask chunking along the spatial dimensions.

I don't know how to measure the data requests though... Is it possible/easy to set up a local OpenDAP server and monitor what comes out?  `memory_profiler` is a good choice for the RAM side of things.

**Comment by agstephens** [comment]
@huard I think the same fix that @ellesmith88 suggested in your other PR should resolve the failing tests here. See:

https://github.com/roocs/clisops/pull/190#issuecomment-929079144

**Comment by huard** [comment]
Using the `memory_profiler` package, I've been able to measure memory usage. 
Profiling code: https://gist.github.com/huard/edfdac4afb5adf1432b0953a1f18e4ce

Without subsetting:
```
Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    17  170.840 MiB  170.840 MiB           1   @profile
    18                                         def avg():
    19  177.195 MiB    6.355 MiB           1       ds = xr.open_dataset(nc_file_neglons)
    20                                         
    21                                             # xESMF has a problem with averaging over dataset when non-averaged variables are present...
    22  239.055 MiB   61.859 MiB           1       average.average_shape(ds.tasmax, small_geojson)
```

With subsetting:
```
Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    17  170.652 MiB  170.652 MiB           1   @profile
    18                                         def avg():
    19  177.168 MiB    6.516 MiB           1       ds = xr.open_dataset(nc_file_neglons)
    20                                         
    21                                             # xESMF has a problem with averaging over dataset when non-averaged variables are present...
    22  212.562 MiB   35.395 MiB           1       average.average_shape(ds.tasmax, small_geojson)
```
So it seems to be working, although not at the magnitude I was hoping for.

**Comment by huard** [comment]
I propose that average_shape only support Dataset inputs, with a new "variable" argument to specify which variable or variables should be averaged over. 
The rationale is that lat_bounds and lon_bounds are not coordinates attached to a DataArray, but variables within the Dataset.

**Comment by huard** [comment]
I'll need help on this one. 
1. There are failing tests in the `conda-xesmf` workflow, but checks are green. 
2. With xesmf 0.6.2 (fresh new release), I'm getting numba errors.

**Comment by aulemahal** [comment]
There's a lot of logging output, I am guessing it's because of #144.

I have only seen the first test and I think it fails only because of your updates to the SpatialAverager, results are different because the handling of the weights is different. Results should be better now, if anything.

File: tests/core/test_subset.py [hunk 1]
```diff
@@ -1056,7 +1056,6 @@ def test_rectilinear(self):
             inds = subset.shape_bbox_indexer(ds, gpd.GeoDataFrame(geometry=[pb]))
             assert pb.within(subset.grid_exterior_polygon(ds.isel(inds)))
 
-    @pytest.mark.skip
```
Comment (aulemahal) [inline tests/core/test_subset.py]: You need `@pytest.mark.skipif(xesmf is None, reason="xESMF is needed for  this test.")`

**Review by aulemahal — COMMENTED** [review]
Line 76 of `.github/workflows/main.yml` says `continue-on-error: true`, wouldn't that be the reason why tests fails but github doesn't stop?

File: tests/core/test_subset.py (hunk 1 above)
Comment (huard) [reply tests/core/test_subset.py]: So you suggest I just update the results with the most recent values ?

Comment (aulemahal) [reply tests/core/test_subset.py]: I think so yeah. You could verify locally that the tests pass with xESMF 0.6.1?

Comment (huard) [reply tests/core/test_subset.py]: Did it with 0.6.0. I still had issues with 0.6.1.

**Review by aulemahal — APPROVED** [review]
Those are nice additions!

**Comment by huard** [comment]
@agstephens Ready for review on your end.

**Comment by agstephens** [comment]
Hi @huard, I am getting an error in `pygeos` at version 0.8:

```
>           pygeos.prepare(pts)
E           AttributeError: module 'pygeos' has no attribute 'prepare'
```

I am trying to update my version of `pygeos`. Which version are you working with?

**Comment by huard** [comment]
0.10.2

**Comment by aulemahal** [comment]
@agstephens  If the error comes from  `core.subset.create_mask`, that might be my error! I wrote this part and added the dependency as `pygeos>=0.8` because that's what `geopandas` says. However, it seems, as your error shows, that I used a function from a later version. Checking the doc, I see that `pygeos.prepare` appears in 0.9, so that's what should have been written the requirements all along!

**Comment by agstephens** [comment]
@huard @aulemahal: Thanks, I have got an update to `pygeos==0.11.1`, it fixes the initial errors, but introduces a different one. I'll try downgrading to `0.10.2` to see if it fixes it. (Might be a problem with my environment so I'll nuke that if 0.10.2 doesn't work).

**Comment by agstephens** [comment]
@huard @aulemahal I am getting the same single error (shown above) now in my rebuilt environment. It's strange that it is a subsetting operation that results in a very slightly different output. I suspect it would be good to know why this is the case. Have you seen that error at all?

**Comment by huard** [comment]
No, will try to reproduce and investigate.

**Comment by huard** [comment]
Can't reproduce, will need info on your environment. 
However, the latest cf-xarray patch release creates new failures. We may have to pin it to 0.6.1.

**Comment by aulemahal** [comment]
There are weird things happening here.

First, I am also unable to replicate the failure seen by @agstephens. I add the output of `conda list` at the end of this comment.

Secondly, @huard  if you are talking about `TestSubsetShaoe::test_wraps`, I am sucessfully able to make it fail with cf_xarray 0.6.1  and 0.6.0.  I believe the issue is that when receiving a `DataArray`, `subset_shape` gets a dataset with `da._to_temp_dataset()` the resulting Dataset has a single variable which has a non-string name. I think this is why `cf_xarray` fails, I'm guessing we could say that's a bug.
But why did it pass on the last build, then? I'm lost. I have the same xarray and cf_xarray versions. Like literally I did:
```python
import xarray as xr
import cf_xarray as cfxr

ds = xr.open_dataset('.../cmip5/tas_Amon_CanESM2_rcp85_r1i1p1_200701-200712.nc')
ds.tas._to_temp_dataset().cf
```
And it fails. It's not an issue of the latest  `cf_xarray`, it 's older than that. Then why do the builds here pass?

<details>

My env.
```
# packages in environment at /opt/miniconda3/envs/clis37:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       1_gnu    conda-forge
alabaster                 0.7.12                   pypi_0    pypi
attrs                     21.2.0             pyhd8ed1ab_0    conda-forge
babel                     2.9.1                    pypi_0    pypi
backcall                  0.2.0                    pypi_0    pypi
backports-entry-points-selectable 1.1.1                    pypi_0    pypi
black                     21.12b0                  pypi_0    pypi
bleach                    4.1.0                    pypi_0    pypi
bokeh                     2.4.2            py37h89c1867_0    conda-forge
boost-cpp                 1.74.0               h359cf19_5    conda-forge
bottleneck                1.3.2            py37hb1e94ed_5    conda-forge
branca                    0.4.2              pyhd8ed1ab_0    conda-forge
brotli                    1.0.9                h7f98852_6    conda-forge
brotli-bin                1.0.9                h7f98852_6    conda-forge
brotlipy                  0.7.0           py37h5e8e339_1003    conda-forge
bump2version              1.0.1                    pypi_0    pypi
bumpversion               0.6.0                    pypi_0    pypi
bzip2                     1.0.8                h7f98852_4    conda-forge
c-ares                    1.18.1               h7f98852_0    conda-forge
ca-certificates           2021.10.8            ha878542_0    conda-forge
cairo                     1.16.0               hf32fb01_1  
certifi                   2021.10.8        py37h89c1867_1    conda-forge
cf-xarray                 0.6.0                    pypi_0    pypi
cffi                      1.14.6           py37hc58025e_0    conda-forge
cfgv                      3.3.1                    pypi_0    pypi
cfitsio                   4.0.0                h9a35b8e_0    conda-forge
cftime                    1.5.1.1          py37hb1e94ed_1    conda-forge
charset-normalizer        2.0.9              pyhd8ed1ab_0    conda-forge
click                     8.0.3            py37h89c1867_1    conda-forge
click-plugins             1.1.1                      py_0    conda-forge
cligj                     0.7.2              pyhd8ed1ab_1    conda-forge
clisops                   0.7.0                     dev_0    <develop>
cloudpickle               2.0.0              pyhd8ed1ab_0    conda-forge
colorama                  0.4.4                    pypi_0    pypi
coverage                  6.2                      pypi_0    pypi
cryptography              36.0.0           py37hf1a17b8_0    conda-forge
curl                      7.80.0               h2574ce0_0    conda-forge
cycler                    0.11.0             pyhd8ed1ab_0    conda-forge
cytoolz                   0.11.2           py37h5e8e339_1    conda-forge
dask                      2021.11.2          pyhd8ed1ab_0    conda-forge
dask-core                 2021.11.2          pyhd8ed1ab_0    conda-forge
decorator                 5.1.0                    pypi_0    pypi
defusedxml                0.7.1                    pypi_0    pypi
distlib                   0.3.4                    pypi_0    pypi
distributed               2021.11.2        py37h89c1867_0    conda-forge
docutils                  0.17.1                   pypi_0    pypi
entrypoints               0.3                      pypi_0    pypi
esmf                      8.2.0           nompi_h61edca3_0    conda-forge
esmpy                     8.2.0           nompi_py37hff05f9c_1    conda-forge
expat                     2.4.1                h9c3ff4c_0    conda-forge
filelock                  3.4.0                    pypi_0    pypi
fiona                     1.8.20           py37h2cf5250_4    conda-forge
flake8                    4.0.1                    pypi_0    pypi
folium                    0.12.1.post1       pyhd8ed1ab_1    conda-forge
font-ttf-dejavu-sans-mono 2.37                 hab24e00_0    conda-forge
font-ttf-inconsolata      3.000                h77eed37_0    conda-forge
font-ttf-source-code-pro  2.038                h77eed37_0    conda-forge
font-ttf-ubuntu           0.83                 hab24e00_0    conda-forge
fontconfig                2.13.1            hba837de_1005    conda-forge
fonts-conda-ecosystem     1                             0    conda-forge
fonts-conda-forge         1                             0    conda-forge
fonttools                 4.28.3           py37h5e8e339_0    conda-forge
freetype                  2.10.4               h0708190_1    conda-forge
freexl                    1.0.6                h7f98852_0    conda-forge
fsspec                    2021.11.1          pyhd8ed1ab_0    conda-forge
gdal                      3.4.0            py37h7b489ff_3    conda-forge
geopandas                 0.10.2             pyhd8ed1ab_1    conda-forge
geopandas-base            0.10.2             pyha770c72_1    conda-forge
geos                      3.10.0               h9c3ff4c_0    conda-forge
geotiff                   1.7.0                h90a4e78_5    conda-forge
gettext                   0.19.8.1          h0b5b191_1005    conda-forge
giflib                    5.2.1                h36c2ea0_2    conda-forge
gitdb                     4.0.9                    pypi_0    pypi
gitpython                 3.1.12                   pypi_0    pypi
glib                      2.68.4               h9c3ff4c_0    conda-forge
glib-tools                2.68.4               h9c3ff4c_0    conda-forge
hdf4                      4.2.15               h10796ff_3    conda-forge
hdf5                      1.12.1          nompi_h2750804_103    conda-forge
heapdict                  1.0.1                      py_0    conda-forge
icu                       69.1                 h9c3ff4c_0    conda-forge
identify                  2.4.0                    pypi_0    pypi
idna                      3.1                pyhd3deb0d_0    conda-forge
imagesize                 1.3.0                    pypi_0    pypi
importlib-metadata        4.2.0                    pypi_0    pypi
importlib-resources       5.4.0                    pypi_0    pypi
importlib_metadata        4.8.2                hd8ed1ab_0    conda-forge
iniconfig                 1.1.1                    pypi_0    pypi
ipython                   7.30.1                   pypi_0    pypi
ipython-genutils          0.2.0                    pypi_0    pypi
jbig                      2.1               h7f98852_2003    conda-forge
jedi                      0.18.1                   pypi_0    pypi
jeepney                   0.7.1                    pypi_0    pypi
jinja2                    3.0.3              pyhd8ed1ab_0    conda-forge
joblib                    1.1.0              pyhd8ed1ab_0    conda-forge
jpeg                      9d                   h36c2ea0_0    conda-forge
json-c                    0.15                 h98cffda_0    conda-forge
jsonschema                4.2.1                    pypi_0    pypi
jupyter-client            7.1.0                    pypi_0    pypi
jupyter-core              4.9.1                    pypi_0    pypi
jupyterlab-pygments       0.1.2                    pypi_0    pypi
kealib                    1.4.14               h87e4c3c_3    conda-forge
keyring                   23.4.0                   pypi_0    pypi
kiwisolver                1.3.2            py37h2527ec5_1    conda-forge
krb5                      1.19.2               hcc1bbae_3    conda-forge
lcms2                     2.12                 hddcbb42_0    conda-forge
ld_impl_linux-64          2.36.1               hea4e1c9_2    conda-forge
lerc                      3.0                  h9c3ff4c_0    conda-forge
libblas                   3.9.0           12_linux64_openblas    conda-forge
libbrotlicommon           1.0.9                h7f98852_6    conda-forge
libbrotlidec              1.0.9                h7f98852_6    conda-forge
libbrotlienc              1.0.9                h7f98852_6    conda-forge
libcblas                  3.9.0           12_linux64_openblas    conda-forge
libcurl                   7.80.0               h2574ce0_0    conda-forge
libdap4                   3.20.6               hd7c4107_2    conda-forge
libdeflate                1.8                  h7f98852_0    conda-forge
libedit                   3.1.20191231         he28a2e2_2    conda-forge
libev                     4.33                 h516909a_1    conda-forge
libffi                    3.3                  h58526e2_2    conda-forge
libgcc-ng                 11.2.0              h1d223b6_11    conda-forge
libgdal                   3.4.0                hc0e84f1_3    conda-forge
libgfortran-ng            11.2.0              h69a702a_11    conda-forge
libgfortran5              11.2.0              h5c6108e_11    conda-forge
libglib                   2.68.4               h3e27bee_0    conda-forge
libgomp                   11.2.0              h1d223b6_11    conda-forge
libiconv                  1.16                 h516909a_0    conda-forge
libkml                    1.3.0             h238a007_1014    conda-forge
liblapack                 3.9.0           12_linux64_openblas    conda-forge
libllvm10                 10.0.1               he513fc3_3    conda-forge
libnetcdf                 4.8.1           nompi_hb3fd0d9_101    conda-forge
libnghttp2                1.43.0               h812cca2_1    conda-forge
libnsl                    2.0.0                h7f98852_0    conda-forge
libopenblas               0.3.18          pthreads_h8fe5266_0    conda-forge
libpng                    1.6.37               h21135ba_2    conda-forge
libpq                     13.5                 hd57d9b9_1    conda-forge
librttopo                 1.1.0                h0ad649c_7    conda-forge
libspatialindex           1.9.3                h9c3ff4c_4    conda-forge
libspatialite             5.0.1               hed11718_11    conda-forge
libssh2                   1.10.0               ha56f1ee_2    conda-forge
libstdcxx-ng              11.2.0              he4da1e4_11    conda-forge
libtiff                   4.3.0                h6f004c6_2    conda-forge
libuuid                   2.32.1            h7f98852_1000    conda-forge
libwebp-base              1.2.1                h7f98852_0    conda-forge
libxcb                    1.14                 h7b6447c_0  
libxml2                   2.9.12               h885dcf4_1    conda-forge
libzip                    1.8.0                h4de3113_1    conda-forge
libzlib                   1.2.11            h36c2ea0_1013    conda-forge
llvmlite                  0.36.0           py37h9d7f4d0_0    conda-forge
locket                    0.2.0                      py_2    conda-forge
lz4-c                     1.9.3                h9c3ff4c_1    conda-forge
mapclassify               2.4.3              pyhd8ed1ab_0    conda-forge
markupsafe                2.0.1            py37h5e8e339_1    conda-forge
matplotlib-base           3.5.0            py37h1058ff1_0    conda-forge
matplotlib-inline         0.1.3                    pypi_0    pypi
mccabe                    0.6.1                    pypi_0    pypi
mistune                   0.8.4                    pypi_0    pypi
msgpack-python            1.0.3            py37h2527ec5_0    conda-forge
munch                     2.5.0                      py_0    conda-forge
munkres                   1.1.4              pyh9f0ad1d_0    conda-forge
mypy-extensions           0.4.3                    pypi_0    pypi
nbclient                  0.5.9                    pypi_0    pypi
nbconvert                 6.3.0                    pypi_0    pypi
nbformat                  5.1.3                    pypi_0    pypi
nbsphinx                  0.8.7                    pypi_0    pypi
ncurses                   6.2                  h58526e2_4    conda-forge
nest-asyncio              1.5.4                    pypi_0    pypi
netcdf-fortran            4.5.3           nompi_h2b6e579_106    conda-forge
netcdf4                   1.5.8           nompi_py37hf784469_101    conda-forge
networkx                  2.6.3              pyhd8ed1ab_1    conda-forge
nodeenv                   1.6.0                    pypi_0    pypi
nspr                      4.32                 h9c3ff4c_1    conda-forge
nss                       3.73                 hb5efdd6_0    conda-forge
numba                     0.53.1           py37hb11d6e1_1    conda-forge
numpy                     1.21.4           py37h31617e3_0    conda-forge
olefile                   0.46               pyh9f0ad1d_1    conda-forge
openjpeg                  2.4.0                hb52868f_1    conda-forge
openssl                   1.1.1l               h7f98852_0    conda-forge
packaging                 21.3               pyhd8ed1ab_0    conda-forge
pandas                    1.3.4            py37he8f5f7f_1    conda-forge
pandocfilters             1.5.0                    pypi_0    pypi
parso                     0.8.3                    pypi_0    pypi
partd                     1.2.0              pyhd8ed1ab_0    conda-forge
pathspec                  0.9.0                    pypi_0    pypi
pcre                      8.45                 h9c3ff4c_0    conda-forge
pexpect                   4.8.0                    pypi_0    pypi
pickleshare               0.7.5                    pypi_0    pypi
pillow                    8.4.0            py37h0f21c89_0    conda-forge
pip                       21.3.1             pyhd8ed1ab_0    conda-forge
pixman                    0.40.0               h36c2ea0_0    conda-forge
pkginfo                   1.8.2                    pypi_0    pypi
platformdirs              2.4.0                    pypi_0    pypi
pluggy                    1.0.0                    pypi_0    pypi
poppler                   21.10.0              ha39eefc_0    conda-forge
poppler-data              0.4.11               hd8ed1ab_0    conda-forge
postgresql                13.5                 h2510834_1    conda-forge
pre-commit                2.16.0                   pypi_0    pypi
proj                      8.2.0                h277dcde_0    conda-forge
prompt-toolkit            3.0.24                   pypi_0    pypi
psutil                    5.8.0            py37h5e8e339_2    conda-forge
ptyprocess                0.7.0                    pypi_0    pypi
py                        1.11.0                   pypi_0    pypi
pycodestyle               2.8.0                    pypi_0    pypi
pycparser                 2.21               pyhd8ed1ab_0    conda-forge
pyflakes                  2.4.0                    pypi_0    pypi
pygeos                    0.11.1           py37ha79c7d7_1    conda-forge
pygments                  2.10.0                   pypi_0    pypi
pyopenssl                 21.0.0             pyhd8ed1ab_0    conda-forge
pyparsing                 3.0.6              pyhd8ed1ab_0    conda-forge
pyproj                    3.2.1            py37hb589d83_5    conda-forge
pyrsistent                0.18.0                   pypi_0    pypi
pysocks                   1.7.1            py37h89c1867_4    conda-forge
pytest                    6.2.5                    pypi_0    pypi
pytest-cov                3.0.0                    pypi_0    pypi
pytest-runner             5.3.1                    pypi_0    pypi
python                    3.7.8           hffdb5ce_3_cpython    conda-forge
python-dateutil           2.8.2              pyhd8ed1ab_0    conda-forge
python_abi                3.7                     2_cp37m    conda-forge
pytz                      2021.3             pyhd8ed1ab_0    conda-forge
pyyaml                    6.0              py37h5e8e339_3    conda-forge
pyzmq                     22.3.0                   pypi_0    pypi
readline                  8.1                  h46c0cb4_0    conda-forge
readme-renderer           31.0                     pypi_0    pypi
requests                  2.26.0             pyhd8ed1ab_1    conda-forge
requests-toolbelt         0.9.1                    pypi_0    pypi
rfc3986                   1.5.0                    pypi_0    pypi
roocs-utils               0.5.0              pyh6c4a22f_0    conda-forge
rtree                     0.9.7            py37h0b55af0_3    conda-forge
scikit-learn              1.0.1            py37hf9e9bfc_2    conda-forge
scipy                     1.7.3            py37hf2a6cf1_0    conda-forge
secretstorage             3.3.1                    pypi_0    pypi
setuptools                59.4.0           py37h89c1867_0    conda-forge
shapely                   1.8.0            py37ha79c7d7_2    conda-forge
six                       1.16.0             pyh6c4a22f_0    conda-forge
smmap                     5.0.0                    pypi_0    pypi
snowballstemmer           2.2.0                    pypi_0    pypi
sortedcontainers          2.4.0              pyhd8ed1ab_0    conda-forge
sparse                    0.13.0             pyhd8ed1ab_0    conda-forge
sphinx                    4.3.1                    pypi_0    pypi
sphinx-rtd-theme          1.0.0                    pypi_0    pypi
sphinxcontrib-applehelp   1.0.2                    pypi_0    pypi
sphinxcontrib-devhelp     1.0.2                    pypi_0    pypi
sphinxcontrib-htmlhelp    2.0.0                    pypi_0    pypi
sphinxcontrib-jsmath      1.0.1                    pypi_0    pypi
sphinxcontrib-qthelp      1.0.3                    pypi_0    pypi
sphinxcontrib-serializinghtml 1.1.5                    pypi_0    pypi
sqlite                    3.37.0               h9cd32fc_0    conda-forge
tblib                     1.7.0              pyhd8ed1ab_0    conda-forge
testpath                  0.5.0                    pypi_0    pypi
threadpoolctl             3.0.0              pyh8a188c0_0    conda-forge
tiledb                    2.3.4                he87e0bf_0    conda-forge
tk                        8.6.11               h27826a3_1    conda-forge
toml                      0.10.2                   pypi_0    pypi
tomli                     1.2.2                    pypi_0    pypi
toolz                     0.11.2             pyhd8ed1ab_0    conda-forge
tornado                   6.1              py37h5e8e339_2    conda-forge
tox                       3.24.4                   pypi_0    pypi
tqdm                      4.62.3                   pypi_0    pypi
traitlets                 5.1.1                    pypi_0    pypi
twine                     3.7.1                    pypi_0    pypi
typed-ast                 1.5.1                    pypi_0    pypi
typing_extensions         4.0.1              pyha770c72_0    conda-forge
tzcode                    2021e                h7f98852_0    conda-forge
tzdata                    2021e                he74cb21_0    conda-forge
unicodedata2              13.0.0.post2     py37h5e8e339_4    conda-forge
urllib3                   1.26.7             pyhd8ed1ab_0    conda-forge
virtualenv                20.10.0                  pypi_0    pypi
watchdog                  2.1.6                    pypi_0    pypi
wcwidth                   0.2.5                    pypi_0    pypi
webencodings              0.5.1                    pypi_0    pypi
wheel                     0.37.0             pyhd8ed1ab_1    conda-forge
xarray                    0.20.2             pyhd8ed1ab_0    conda-forge
xerces-c                  3.2.3                h8ce2273_4    conda-forge
xesmf                     0.6.2              pyhd8ed1ab_0    conda-forge
xyzservices               2021.11.0          pyhd8ed1ab_1    conda-forge
xz                        5.2.5                h516909a_1    conda-forge
yaml                      0.2.5                h516909a_0    conda-forge
zict                      2.0.0                      py_0    conda-forge
zipp                      3.6.0              pyhd8ed1ab_0    conda-forge
zlib                      1.2.11            h36c2ea0_1013    conda-forge
zstd                      1.5.0                ha95c52a_0    conda-forge
```

</details>

**Comment by huard** [comment]
@aulemahal yes, but weird, because here I fixed the error by downgrading cfxr. I suspect your env. still finds the latest cfxr somewhere (are you using pycharm to run the tests ?)

**Comment by aulemahal** [comment]
Ayoye, this gets weirder and weirder!

I created a fully new env with `cf_xarray` 0.6.0 and the following code fails:
```
import xarray as xr
import cf_xarray as cfxr

ds = xr.Dataset(
    {'a': (('x',), [1, 2, 3, 4])},
    coords={
        'x': [2, 3, 4, 5],
        'x_bnds': (('bnds', 'x'), [[1.5, 2.5, 3.5, 4.5], [2.5, 3.5, 4.5, 5.5]])
    }
)
ds.x.attrs['bounds'] = 'x_bnds'

print(cfxr.__version__)
print(ds.a._to_temp_dataset().cf)
 ```
I get the same error as when `test_wraps` fails on my side. 

Moving to `cf_xarray` 0.6.2 did not resolve the issue.

So, I think I found a `cf_xarray` bug that exists in `subset_shape` in at least a few cases. There might be something else going on that makes the bug disappear on the github builds. May be I stumbled on this by chance and it's not the same bug as your seeing? 

I'll open a ticket on `cf_xarray` and continue my descent.

**Comment by agstephens** [comment]
Hi folks, sorry for lack of responses, I've been ill this week. I have tried pulling your changes and I still get the same error. I'm trying to modify my environment to match yours @aulemahal in the hope that it will show the package causing the error.

**Comment by agstephens** [comment]
@huard @aulemahal , even when I try to reproduce your environment above - I still get the same single error. I'm very confused!

**Comment by agstephens** [comment]
Hi @huard @aulemahal, this time I installed a brand new VM to run the tests - with an environment unpolluted by anything else that might be causing side effects.
And all the tests passed! I'm merging.

**Comment by huard** [comment]
Thanks !

**Pull request merged**